core/src/mast/serialization/mod.rs

```diff
@@ -0,0 +1,165 @@
+    }
+}
+
+// MAST FOREST SERIALIZATION/DESERIALIZATION
+// ===============================================================================================
+
+impl Serializable for MastForest {
+    fn write_into<W: ByteWriter>(&self, target: &mut W) {
+        let mut basic_block_data_builder = BasicBlockDataBuilder::new();
+
+        // magic & version
+        target.write_bytes(MAGIC);
+        target.write_bytes(&VERSION);
+
+        // node count
+        target.write_usize(self.nodes.len());
+
+        // roots
+        self.roots.write_into(target);
+
+        // MAST node infos
+        for mast_node in &self.nodes {
+            let mast_node_info =
+                MastNodeInfo::new(mast_node, basic_block_data_builder.next_data_offset());
+
+            if let MastNode::Block(basic_block) = mast_node {
+                basic_block_data_builder.encode_basic_block(basic_block);
+            }
+
+            mast_node_info.write_into(target);
+        }
+
+        let (data, string_table) = basic_block_data_builder.into_parts();
+
+        string_table.write_into(target);
+        data.write_into(target);
+    }
+}
+
+impl Deserializable for MastForest {
+    fn read_from<R: ByteReader>(source: &mut R) -> Result<Self, DeserializationError> {
+        let magic: [u8; 5] = source.read_array()?;
+        if magic != *MAGIC {
+            return Err(DeserializationError::InvalidValue(format!(
+                "Invalid magic bytes. Expected '{:?}', got '{:?}'",
+                *MAGIC, magic
+            )));
+        }
+
+        let version: [u8; 3] = source.read_array()?;
+        if version != VERSION {
+            return Err(DeserializationError::InvalidValue(format!(
+                "Unsupported version. Got '{version:?}', but only '{VERSION:?}' is supported",
+            )));
+        }
+
+        let node_count = source.read_usize()?;
+
+        let roots: Vec<MastNodeId> = Deserializable::read_from(source)?;
+
+        let mast_node_infos = {
+            let mut mast_node_infos = Vec::with_capacity(node_count);
+            for _ in 0..node_count {
+                let mast_node_info = MastNodeInfo::read_from(source)?;
+                mast_node_infos.push(mast_node_info);
+            }
+
+            mast_node_infos
+        };
+
+        let strings: Vec<StringRef> = Deserializable::read_from(source)?;
+
+        let data: Vec<u8> = Deserializable::read_from(source)?;
+
+        let mut basic_block_data_decoder = BasicBlockDataDecoder::new(&data, &strings);
+
+        let mast_forest = {
+            let mut mast_forest = MastForest::new();
+
+            for mast_node_info in mast_node_infos {
+                let node = mast_node_info
+                    .try_into_mast_node(&mast_forest, &mut basic_block_data_decoder)?;
+
+                mast_forest.add_node(node);
+            }
+
+            for root in roots {
+                mast_forest.make_root(root);
+            }
+
+            mast_forest
+        };
```

Comment:
Seems like we are iterating over node infos twice here: first time as we are deserializing them and second time when we try to deserialize the associated data. I think if we altered how we serialize things a bit, we could avoid this. Specifically:

Currently, we have the following order:
```
[magic, version, num_nodes, roots, node_infos, string_table, data]
```
But maybe the following is better:
```
[magic, version, num_nodes, roots, string_table, data, node_infos]
```
This way, we deserialize string table and data first and so we have everything in memory and would need to iterate over node infos only once.

This will make serialization a bit less efficient - but I think we care more about deserialization speed then serialization speed.